### DISCUSSION TITLE
[RFC] Classifier state and feedback, interventions or similar

### DISCUSSION THREAD

**Discussion by mcbouslog** [zooniverse/front-end-monorepo#2423] [General]
**Package**
lib-classifier

**Feature or Issue Description**
Per #549 to show the feedback modal we use [MST middleware](https://github.com/mobxjs/mobx-state-tree/blob/master/docs/middleware.md) to catch the action to advance to the next subject, which we then abort. The feedback modal is shown, and on close we reconnect to the subjects store and advance to the next subject. We're not wild about this implementation.

**Why we're not wild about the current implementation:**
- [per this PR comment](https://github.com/zooniverse/front-end-monorepo/pull/549#discussion_r267436080): One of my dislikes about this implementation is that it's recursive. Calling `subjects.advance()` at the end of feedback will trigger this handler to run again. We've hacked around that by checking to see if there's a modal open. It's possible to accidentally put the code into a never-ending loop where closing the feedback modal opens it again and you never get to the next subject.
- paraphrased from other PR comments: It'd be nice if we could use a pattern where feedback interrupts the action in progress, then resumes it. At the moment, it just blocks the subject queue from advancing and we have to remember to advance it again somewhere else in the code. ~~At the moment, the feedback modal will always advance the subject queue on close, no matter where it's used in the classifier.~~
- per Slack: Ideally, this should pause the action, show feedback (or whichever intervention is happening) then continue the paused action. That will become more important when we implement interventions in the new classifier.

**Let's use this RFC to discuss potential refactors**
- an overarching classifier state store might be useful, to pause/resume the classifier flow, which feedback, interventions or other features could hook into that the rest of the classification process would react to.

**Comment by eatyourgreens** [comment]
The feedback modal isn’t hardwired to advance the subject queue, as of today’s changes.

**Comment by eatyourgreens** [comment]
A couple of issues with the current implementation:

- Recursion: we listen for `subjects.advance`, then show a modal which fires `subject.advance` on close, which checks to show feedback etc. etc. Change the order of the following two lines of code and the feedback modal will always re-open on `hideFeedback`. https://github.com/zooniverse/front-end-monorepo/blob/405f797f85557320fef5c288c5e07f07288f95ea/packages/lib-classifier/src/store/FeedbackStore.js#L107-L110

- Going to Talk bypasses feedback. Done & Talk (#516) redirects the window to a new URL before feedback can open. In PFE, we fixed this by hijacking Done & Talk and redirecting after feedback is shown. https://github.com/zooniverse/Panoptes-Front-End/pull/4450#issuecomment-373089470

**Comment by eatyourgreens** [comment]
Looking back at PFE, the Done button triggers showing feedback, which then triggers `completeClassification` when closed. The original click event is passed through to `completeClassification` which then branches based on whether the clicked element was a link (Done & Talk) or a button (Done.)

**Comment by srallen** [comment]
That's pretty ugly and another reason to remove the means to go to Talk from the task navigation.

**Comment by eatyourgreens** [comment]
For TESS, I'm thinking Done/Done & Talk triggers `showFeedback`, which either shows feedback or passes through to finishing the classification. Done & Talk would need to store something that tells us whether the next action after completing the classification is opening Talk or advancing to the next subject.

**Comment by eatyourgreens** [comment]
And we're going to have to deal with this once #516 merges.

**Comment by eatyourgreens** [comment]
I've opened the Done & Talk issue as #598.
